plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1754 @@
+            error (show trm ++ " " ++ show type)
+          else
+            return (trm, type)
+    ```
+-}
+
+-- | Term generators carry around a context to know
+-- e.g. what types and terms are in scope.
+type GenTm = ReaderT GenEnv Gen
+
+data GenEnv = GenEnv
+  { geSize               :: Int
+  -- ^ Generator size bound
+  , geDatas              :: Map TyName (Datatype TyName Name DefaultUni DefaultFun ())
+  -- ^ Datatype context
+  , geTypes              :: Map TyName (Kind ())
+  -- ^ Type context
+  , geTerms              :: Map Name (Type TyName DefaultUni ())
+  -- ^ Term context
+  , geUnboundUsedTyNames :: Set TyName
+  -- ^ Names that we have generated and don't want to shadow but haven't bound yet.
+  , geEscaping           :: AllowEscape
+  -- ^ Are we in a place where we are allowed to generate a datatype binding?
+  , geCustomGen          :: Maybe (Type TyName DefaultUni ())
+                         -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ())
+  -- ^ A custom user-controlled generator for terms - useful for situations where
+  -- we want to e.g. generate custom strings for coverage or test some specific
+  -- pattern that generates a special case for the compiler.
+  , geCustomFreq         :: Int
+  -- ^ How often do we use the custom generator -
+  -- values in the range of 10-30 are usually reasonable.
+  }
+
+-- | Run a `GenTm  generator in a top-level empty context where we are allowed to generate
+-- datatypes.
+runGenTm :: GenTm a -> Gen a
+runGenTm = runGenTmCustom 0 (error "No custom generator.")
+
+-- | Run a `GenTm` generator with a plug-in custom generator for terms that is included with
+-- the other generators.
+runGenTmCustom :: Int
+               -> (Maybe (Type TyName DefaultUni ())
+                  -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ()))
+               -> GenTm a
+               -> Gen a
+runGenTmCustom f cg g = sized $ \ n ->
+  runReaderT g $ GenEnv { geSize               = n
+                        , geDatas              = Map.empty
+                        , geTypes              = Map.empty
+                        , geTerms              = Map.empty
+                        , geUnboundUsedTyNames = Set.empty
+                        , geEscaping           = YesEscape
+                        , geCustomGen          = cg
+                        , geCustomFreq         = f
+                        }
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+noEscape :: GenTm a -> GenTm a
+noEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Functions for lifting `Gen` stuff to `GenTm`
+
+-- | Lift `Gen` generator to `GenTm` generator. Respects `geSize`.
+liftGen :: Gen a -> GenTm a
+liftGen gen = do
+  sz <- asks geSize
+  lift $ resize sz gen
+
+-- | Lift functor operations like `oneof` from `Gen` to `GenTm`
+liftGenF :: Functor f => (f (Gen a) -> Gen a) -> f (GenTm a) -> GenTm a
```

Comment:
I think something like that would make a lot of this *way* easier.

Reply:
Generally, there's a lot of general "monad transformer on top of Gen" wrangling here that seems like it could live in a separate module, or ideally in `Quickcheck` itself. This stuff is pretty painless with hedgehog ;)

Reply:
Would it be possible to just use [`QuickCheck-GenT`](https://hackage.haskell.org/package/QuickCheck-GenT-0.2.2/docs/QuickCheck-GenT.html)?

Reply:
No, because it puts the reader bits in the wrong place - all the non-determinism is *outside* the parameter space.

Reply:
Sorry, could you clarify? `Gen` from `QuickCheck` is itself a reader monad (by nature, not literally) and two readers commute perfectly well, so the order shouldn't matter. `GenT` lacks `MonadReader`, `MonadState` etc instances, but it's trivial to provide them, e.g.

```haskell
instance MonadState s m => MonadState s (GenT m) where
    get = lift get
    put = lift . put
    state = lift . state
```